## Merging in Python
- Merging in python is the process of combining two or more DataFrames which are tables into one based on a common column.
- Merging is important because it allows us to bring related data from different tables into one dataset, it adds context to raw data by adding meaningful attributes to it.It also makes data ready for analysis and reporting.

In [13]:
import pandas as pd

In [14]:
file_path="PowerBI Customers Sales tables.xlsx"

In [30]:
customer_data= pd.read_excel(file_path,sheet_name=None)

In [ ]:
print(customer_data.keys())# will show the sheets name in a dictionary way showing the keys

dict_keys(['Customers', 'Products', 'Sales', 'Calendar', 'Regions', 'Categories'])


In [34]:
# using loop to see the first five rows of each sheet
for sheet_name, df in customer_data.items():
    print(f"\n--- {sheet_name} ---")
    print(df.head())


--- Customers ---
   CustomerID CustomerName   Region  Region id
0         101        Alice  Nairobi          1
1         102          Bob   Nakuru          2
2         103      Charlie  Mombasa          3
3         104        Diana  Nairobi          1
4         105        Ethan   Kisumu          4

--- Products ---
   ProductID ProductName Category
0        201       Maize   Grains
1        202       Beans   Grains
2        203        Rice   Grains
3        204       Wheat   Grains
4        205     Sorghum   Grains

--- Sales ---
   SaleID  CustomerID  ProductID  Quantity  UnitPrice   SaleDate  Revenue
0       1         101        201         5         50 2023-01-01      250
1       2         102        202         3         60 2023-02-01      180
2       3         103        203         8         70 2023-03-01      560
3       4         104        204         2         55 2023-04-01      110
4       5         101        205         4         45 2023-05-01      180

--- Calendar ---


In [35]:
#to see the columns of the sheets
for sheet_name, df in customer_data.items():
    print(f"\n--- {sheet_name} ---")
    print(df.columns.tolist())


--- Customers ---
['CustomerID', 'CustomerName', 'Region', 'Region id']

--- Products ---
['ProductID', 'ProductName', 'Category']

--- Sales ---
['SaleID', 'CustomerID', 'ProductID', 'Quantity', 'UnitPrice', 'SaleDate', 'Revenue']

--- Calendar ---
['Date', 'Month', 'Quarter', 'Year']

--- Regions ---
['Region ID', 'Region', 'RegionManager']

--- Categories ---
['Category id', 'Category', 'CategoryDescription']


In [36]:
#to see the summary 
for sheet_name, df in customer_data.items():
    print(f"\n--- {sheet_name} ---")
    print(df.info())


--- Customers ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   CustomerID    10 non-null     int64 
 1   CustomerName  10 non-null     object
 2   Region        10 non-null     object
 3   Region id     10 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 452.0+ bytes
None

--- Products ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ProductID    11 non-null     int64 
 1   ProductName  11 non-null     object
 2   Category     11 non-null     object
dtypes: int64(1), object(2)
memory usage: 396.0+ bytes
None

--- Sales ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      -

In [37]:
#extract tables this helps give structure,control and flexibility
customers=customer_data["Customers"]
products=customer_data["Products"]
sales=customer_data["Sales"]
calendar=customer_data["Calendar"]
regions=customer_data["Regions"]
categories=customer_data["Categories"]

In [52]:
categories

Category id    Category                                CategoryDescription
0            1      Grains            Dry food grains like maize, wheat, etc.
1            2  Vegetables  Fresh farm vegetables like tomatoes, cabbage, ...

In [51]:
regions

Region ID   Region RegionManager
0          1  Nairobi         Grace
1          2   Nakuru         Henry
2          3  Mombasa         Irene
3          4   Kisumu          Jack
4          5    Narok           Dan

In [50]:
calendar

Date Month Quarter  Year
0  2023-01-01   Jan      Q1  2023
1  2023-02-01   Feb      Q1  2023
2  2023-03-01   Mar      Q1  2023
3  2023-04-01   Apr      Q2  2023
4  2023-05-01   May      Q2  2023
5  2023-06-01   Jun      Q2  2023
6  2023-07-01   Jul      Q3  2023
7  2023-08-01   Aug      Q3  2023
8  2023-09-01   Sep      Q3  2023
9  2023-10-01   Oct      Q4  2023
10 2023-11-01   Nov      Q4  2023
11 2023-12-01   Dec      Q4  2023

In [49]:
products

ProductID ProductName    Category
0         201       Maize      Grains
1         202       Beans      Grains
2         203        Rice      Grains
3         204       Wheat      Grains
4         205     Sorghum      Grains
5         206      Barley      Grains
6         207    Potatoes  Vegetables
7         208    Tomatoes  Vegetables
8         209      Onions  Vegetables
9         210     Cabbage  Vegetables
10        123        Peas      Grains

In [40]:
customers

CustomerID CustomerName   Region  Region id
0         101        Alice  Nairobi          1
1         102          Bob   Nakuru          2
2         103      Charlie  Mombasa          3
3         104        Diana  Nairobi          1
4         105        Ethan   Kisumu          4
5         106        Fiona   Nakuru          2
6         107       George  Mombasa          3
7         108       Hannah   Kisumu          4
8         109          Ian  Nairobi          1
9         110          Dan    Narok          5

In [41]:
sales

SaleID  CustomerID  ProductID  Quantity  UnitPrice   SaleDate  Revenue
0        1         101        201         5         50 2023-01-01      250
1        2         102        202         3         60 2023-02-01      180
2        3         103        203         8         70 2023-03-01      560
3        4         104        204         2         55 2023-04-01      110
4        5         101        205         4         45 2023-05-01      180
5        6         106        206         7         65 2023-06-01      455
6        7         107        207         6         30 2023-07-01      180
7        8         108        208         1         40 2023-08-01       40
8        9         101        209         9         35 2023-09-01      315
9       10         110        210        10         25 2023-10-01      250
10      11         134        234        10        536 2023-10-01      656

## Knowing what to merge
- Identify the fact table, a fact table contains transactions or events(what actually happened) and is usually the biggest table.
Example a customerID 101 bought a product with ID 201 on 1st Jan 2023.
    - the sales table is the fact table it has transactions that is the salesID, customerID,productID 
- Identify the demension tables, a demension table describe attributes.(describes what happened)
    - Adds attributes example customerID 101 ="Alice",Region="Nairobi".productID 201 is "Maize", category="Grains"
- Dimensions:

Customers (join on CustomerID)

Products (join on ProductID)

Calendar (join SaleDate ↔ Date)

Regions (join Region id ↔ Region ID)

Categories (join on Category text)    

Analytics/reporting/Fact+ dimension → LEFT JOIN

Matching only/ for comparing or filtering datasets → INNER JOIN

Coverage checks → OUTER JOIN 
- eg when you want to check which customers existin the customers table but never appear in sales 

In [ ]:
# Merge Sales with Customers using the left join
merged = sales.merge(customers, on="CustomerID", how="left")

In [43]:
# Merge with Products using the left join
merged = merged.merge(products, on="ProductID", how="left")

In [ ]:
# Merge with Calendar takes the saleDate column from sales and matches it with the date column in the calender table, left join because it kepps all sales 
merged = merged.merge(calendar, left_on="SaleDate", right_on="Date", how="left")

In [ ]:
# Merge Customers → Regions, 
# left_on uses the column with Region id in the merged dataframe which comes from customers
#right_on uses the column name Region ID in the Regions table 
# we can't use on because of id and ID 
merged = merged.merge(regions, left_on="Region id", right_on="Region ID", how="left")

In [ ]:
# Merge Products → Categories
merged = merged.merge(categories, left_on="Category", right_on="Category", how="left")

In [47]:
print(merged.head())

   SaleID  CustomerID  ProductID  Quantity  UnitPrice   SaleDate  Revenue  \
0       1         101        201         5         50 2023-01-01      250   
1       2         102        202         3         60 2023-02-01      180   
2       3         103        203         8         70 2023-03-01      560   
3       4         104        204         2         55 2023-04-01      110   
4       5         101        205         4         45 2023-05-01      180   

  CustomerName Region_x  Region id  ... Category       Date Month Quarter  \
0        Alice  Nairobi        1.0  ...   Grains 2023-01-01   Jan      Q1   
1          Bob   Nakuru        2.0  ...   Grains 2023-02-01   Feb      Q1   
2      Charlie  Mombasa        3.0  ...   Grains 2023-03-01   Mar      Q1   
3        Diana  Nairobi        1.0  ...   Grains 2023-04-01   Apr      Q2   
4        Alice  Nairobi        1.0  ...   Grains 2023-05-01   May      Q2   

   Year  Region ID  Region_y RegionManager Category id  \
0  2023        1